File: examples/cpp/minimal-order-service/domain/money.cpp
#include "domain/money.hpp"
#include <algorithm>
#include <cctype>

namespace domain {

Money::Money(int64_t amount, const std::string& currency)
    : amount_(amount), currency_(currency) {
    if (amount < 0) {
        throw std::invalid_argument("Amount cannot be negative");
    }
    if (currency.empty() || std::all_of(currency.begin(), currency.end(), ::isspace)) {
        throw std::invalid_argument("Currency cannot be empty");
    }
    
    // Normalize currency to uppercase
    currency_.clear();
    for (char c : currency) {
        if (!std::isspace(c)) {
            currency_ += static_cast<char>(std::toupper(c));
        }
    }
}

Money Money::add(const Money& other) const {
    if (currency_ != other.currency_) {
        throw std::invalid_argument("Cannot add different currencies");
    }
    return Money(amount_ + other.amount_, currency_);
}

} // namespace domain

File: examples/go/minimal-order-service/domain/money.go
package domain

import (
	"errors"
	"strings"
)

// Money represents a monetary value.
// It's a value object - immutable and validated.
type Money struct {
	amount   int64  // in cents
	currency string
}

// NewMoney creates a new Money value object.
func NewMoney(amount int64, currency string) (*Money, error) {
	if amount < 0 {
		return nil, errors.New("amount cannot be negative")
	}
	if strings.TrimSpace(currency) == "" {
		return nil, errors.New("currency cannot be empty")
	}
	return &Money{
		amount:   amount,
		currency: strings.ToUpper(strings.TrimSpace(currency)),
	}, nil
}

// Amount returns the amount in cents.
func (m *Money) Amount() int64 {
	return m.amount
}

// Currency returns the currency code.
func (m *Money) Currency() string {
	return m.currency
}

// Add adds two Money values with the same currency.
func (m *Money) Add(other *Money) (*Money, error) {
	if m.currency != other.currency {
		return nil, errors.New("cannot add different currencies")
	}
	return NewMoney(m.amount+other.amount, m.currency)
}

// Equals checks if two Money values are equal.
func (m *Money) Equals(other *Money) bool {
	if other == nil {
		return false
	}
	return m.amount == other.amount && m.currency == other.currency
}
